hi-ml-histopathology/src/histopathology/models/deepmil.py

```diff
@@ -352,20 +352,95 @@ def get_bag_label(labels: Tensor) -> Tensor:
         # SlidesDataModule attributes a single label to a bag of tiles already no need to do majority voting
         return labels
 
+    @staticmethod
+    def get_empty_lists(shape: int, n: int) -> List:
+        ll = []
+        for _ in range(n):
+            ll.append([None] * shape)
+        return ll
+
+    @staticmethod
+    def get_patch_coordinate(slide_offset: List, patch_location: List, patch_size: List) -> Tuple[int, int, int, int]:
+        """ computing absolute patch coordinate """
+        #  PATCH_LOCATION is expected to have shape [y, x]
+        top = slide_offset[0] + patch_location[0]
+        bottom = slide_offset[0] + patch_location[0] + patch_size[0]
+        left = slide_offset[1] + patch_location[1]
+        right = slide_offset[1] + patch_location[1] + patch_size[1]
+        return top, bottom, left, right
+
+    @staticmethod
+    def expand_slide_constant_metadata(id: str, path: str, n_patches: int) -> Tuple[List, List, List]:
+        """Duplicate metadata that is patch invariant to match the shape of other arrays"""
+        slide_id = [id] * n_patches
+        image_paths = [path] * n_patches
+        tile_id = [f"{id}_{tile_id}" for tile_id in range(n_patches)]
+        return slide_id, image_paths, tile_id
+
+    @staticmethod
+    def check_patch_location_format(batch):
+        """Workaround for bug in MONAI that returns not consistent location"""
+        faulty_slides_idx = []
+        for i, locations in enumerate(batch[SlideKey.PATCH_LOCATION]):
+            for location in locations:
+                if len(location) != 2:
+                    print(f'Slide {batch[SlideKey.SLIDE_ID][i]} '
```

Comment:
can you use logging.warninig instead of print?

Reply:
Method no longer needed - patch to MONAI now in dev